azure-iot-device/azure/iot/device/common/mqtt_transport.py

```diff
@@ -463,6 +463,13 @@ def disconnect(self):
             # to this error.
             err = _create_error_from_rc_code(rc)
             raise err
+        else:
+            # Clear pending ops if instructed, but only if the disconnect was successful.
```

Comment:
>but only if the disconnect was successful. [](http://example.com/codeflow?start=47&length=42)

I don't agree with the "don't cancel ops if disconnect fails" behavior.

This is because I have an ICM where an internal Paho socket is getting broken and it's causing mqtt_c;lient.disconnect to fail.  

We need to stop thinking about "disconnect failed" as "we don't know thte state.  We might still be connected" and instead treat it as "Paho is completely messed up and we need to clean up as best as possible."

This ICM goes beyond the change that you're making since the `except Exception as e` line above (448) is where the problem in my ICM manifests.  I'm only mentioning it 
so we can start changing our thinking.

Reply:
Okay, I'll leave this for now, but this is good to consider.